ouroboros-consensus/tools/db-convert/Main.hs

```diff
@@ -0,0 +1,275 @@
+  | InitialChainStateError Scheduling.Error
+  deriving (Show, Typeable)
+
+instance Exception ValidationError
+
+main :: IO ()
+main = do
+  (cmd :: Args Unwrapped) <- unwrapRecord "Byron DB converter"
+  case cmd of
+    Convert {epochDir, dbDir, epochSlots} -> do
+      (_, files) <- listDir $ epochDir
+      let epochFiles = filter (\f -> fileExtension f == ".epoch") files
+      putStrLn $ "Writing to " ++ show dbDir
+      for_ epochFiles $ \f -> do
+        putStrLn $ "Converting file " ++ show f
+        convertEpochFile (EpochSlots epochSlots) f dbDir
+    Validate
+      { dbDir
+      , configFile
+      , requiresNetworkMagic
+      , genesisHash
+      , verbose
+      } -> do
+        econfig <-
+          runExceptT $
+            CC.Genesis.mkConfigFromFile
+              (if requiresNetworkMagic then RequiresMagic else RequiresNoMagic)
+              (toFilePath configFile)
+              genesisHash
+        case econfig of
+          Left err -> throwIO $ MkConfigError err
+          Right config -> validateChainDb dbDir config verbose
+
+convertEpochFile
+  :: EpochSlots
+  -> Path Abs File -- ^ Input
+  -> Path Abs Dir -- ^ Ouput directory
+  -> IO (Either CC.ParseError ())
+convertEpochFile es inFile outDir =
+  let inStream = CC.parseEpochFileWithBoundary es (toFilePath inFile)
+      dbDir = outDir </> [reldir|immutable|]
+      encode = CB.serializeEncoding' . Byron.encodeByronBlock . Byron.ByronBlockOrEBB
+   in do
+        createDirIfMissing True dbDir
+        outFileName <-
+          parseRelFile $ "epoch-" <>
+            drop 2 (toFilePath (filename inFile))
+        outFile <- (dbDir </> outFileName) -<.> "dat"
+        IO.withFile (toFilePath outFile) IO.WriteMode $ \h ->
+          runResourceT $ runExceptT $ S.mapM_ (liftIO . BS.hPut h) . S.map encode $ inStream
+
+validateChainDb
+  :: forall blk. (blk ~ ByronBlockOrEBB ByronConfig)
+  => Path Abs Dir -- ^ DB directory
+  -> CC.Genesis.Config
+  -> Bool -- Verbose
+  -> IO ()
+validateChainDb dbDir cfg verbose =
+  do
+    initialCVS <- runExceptT $ CC.initialChainValidationState cfg
+    case initialCVS of
+      Left err -> throwIO $ InitialChainStateError err
+      Right cvs ->
+        withThreadRegistry $ \registry -> do
+          chaindb <- give protocolMagicId $ give epochSlots $ ChainDB.openDB $ args registry
+          blk <- ChainDB.getTipBlock chaindb
+          putStrLn $ "DB tip: " ++ condense blk
+          ChainDB.closeDB chaindb
+        where
+          initLedgerState = ExtLedgerState
+            { ledgerState = Byron.ByronEBBLedgerState $
+                Byron.ByronLedgerState cvs Seq.empty
+            , ouroborosChainState = Seq.empty
+            }
+          protocolMagicId = CB.unAnnotated . getAProtocolMagicId $ CC.Genesis.configProtocolMagic cfg
+          epochSlots = CC.Genesis.configEpochSlots cfg
+          securityParam = SecurityParam $ CC.unBlockCount k
+          k = CC.Genesis.configK cfg
+          args registry =
```

Comment:
Using a a few helpers (just in this `where` block is fine) would make it more readable, e.g., a helper for making the `args`, a helper for making the `NodeConfig` (separate one for `PBftNodeConfig` and `ByronConfig`?), a helper for the `Tracer`.